File: 933 Number of Recent Calls.py
# Time: O(n)
# Space: O(n)
from collections import deque

class RecentCounter(object):

    def __init__(self):
        # initialize a queue to store recent ping timestamps
       self.recentRequests = deque() 

    def ping(self, t):
        """
        :type t: int
        :rtype: int
        """
        # append current ping timestamp
        self.recentRequests.append(t)

        # remove invalid ping timestamps older than t - 3000
        while self.recentRequests[0] < (t - 3000):
            self.recentRequests.popleft()
        
        # return the number of valid ping timestamps in the queue
        return len(self.recentRequests)


# Your RecentCounter object will be instantiated and called as such:
# obj = RecentCounter()
# param_1 = obj.ping(t)

File: 933 Number of Recent Calls.cpp
// Time: O(n)
// Space: O(n)

#include <queue>

using namespace std;

class RecentCounter {
public:
    RecentCounter() {
        // set the queue to empty
        recentRequests = {}; 
    }
    
    int ping(int t) {
        // push current timestamp to queue and determine the oldest valid timestamp
        recentRequests.push(t);
        int oldest = t - 3000;

        // remove any timestamp older than the oldest valid timestamp
        while (!recentRequests.empty() && recentRequests.front() < oldest){
            recentRequests.pop();
        }

        // return number of valid timestamps in the queue
        return recentRequests.size();
    }

private:
    // a queue data structure to store recent pings
    queue<int> recentRequests;
};

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter* obj = new RecentCounter();
 * int param_1 = obj->ping(t);
 */
